src/mcp/tools/core/validate_rules.ts

```diff
@@ -0,0 +1,149 @@
+import { z } from "zod";
+import { tool } from "../../tool";
+import { mcpError, toContent } from "../../util";
+import { testRuleset } from "../../../gcp/rules";
+import { resolve } from "path";
+import { Client } from "../../../apiv2";
+import { updateRulesWithClient } from "../../../rtdb";
+import { getErrMsg } from "../../../error";
+
+interface SourcePosition {
+  fileName?: string;
+  line?: number;
+  column?: number;
+  currentOffset?: number;
+  endOffset?: number;
+}
+
+interface Issue {
+  sourcePosition: SourcePosition;
+  description: string;
+  severity: string;
+}
+
+function formatRulesetIssues(issues: Issue[], rulesSource: string): string {
+  const sourceLines = rulesSource.split("\n");
+  const formattedOutput: string[] = [];
+
+  for (const issue of issues) {
+    const { sourcePosition, description, severity } = issue;
+
+    let issueString = `${severity}: ${description} [Ln ${sourcePosition.line}, Col ${sourcePosition.column}]`;
+
+    if (sourcePosition.line) {
+      const lineIndex = sourcePosition.line - 1;
+      if (lineIndex >= 0 && lineIndex < sourceLines.length) {
+        const errorLine = sourceLines[lineIndex];
+        issueString += `\n\`\`\`\n${errorLine}`;
+
+        if (
+          sourcePosition.column &&
+          sourcePosition.currentOffset &&
+          sourcePosition.endOffset &&
+          sourcePosition.column > 0 &&
+          sourcePosition.endOffset > sourcePosition.currentOffset
+        ) {
+          const startColumnOnLine = sourcePosition.column - 1;
+          const errorTokenLength = sourcePosition.endOffset - sourcePosition.currentOffset;
+
+          if (
+            startColumnOnLine >= 0 &&
+            errorTokenLength > 0 &&
+            startColumnOnLine <= errorLine.length
+          ) {
+            const padding = " ".repeat(startColumnOnLine);
+            const carets = "^".repeat(errorTokenLength);
+            issueString += `\n${padding}${carets}\n\`\`\``;
+          }
+        }
+      }
+    }
+    formattedOutput.push(issueString);
+  }
+  return formattedOutput.join("\n\n");
+}
+
+export const validate_rules = tool(
+  {
+    name: "validate_rules",
+    description: "Checks the provided Firebase Rules source for syntax and validation errors.",
+    inputSchema: z.object({
+      type: z.enum(["firestore", "storage", "rtdb"]),
+      source: z
+        .string()
+        .optional()
+        .describe("The rules source code to check. Provide either this or a path."),
+      source_file: z
+        .string()
+        .optional()
+        .describe(
+          "A file path, relative to the project root, to a file containing the rules source you want to validate. Provide this or source, not both.",
+        ),
+      databaseUrl: z
+        .string()
+        .optional()
+        .describe(
+          "For RTDB, connect to the database at url. If omitted, use default database instance. Can point to emulator URL.",
+        ),
```

Comment:
Couldn't this apply to Firestore and Storage as well (multiple DBs / multiple buckets)? I think we probably either make this a general `resource_id` which is RTDB instance name, firestore database name, or bucket name, or we drop this altogether and do "defaults only" for now.